pkg/auth/providers/migration/activedirectory_migration.go

```diff
@@ -0,0 +1,243 @@
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	"k8s.io/client-go/tools/cache"
+)
+
+const (
+	crtbsByPrincipalAndUserIndex = "auth.management.cattle.io/crtbByPrincipalAndUser"
+	prtbsByPrincipalAndUserIndex = "auth.management.cattle.io/prtbByPrincipalAndUser"
+	tokenByUserRefKey            = "auth.management.cattle.io/token-by-user-ref"
+)
+
+type adMigration struct {
+	users        v3.UserInterface
+	userLister   v3.UserLister
+	tokenLister  v3.TokenLister
+	tokens       v3.TokenInterface
+	prtbs        v3.ProjectRoleTemplateBindingInterface
+	crtbs        v3.ClusterRoleTemplateBindingInterface
+	crtbIndexer  cache.Indexer
+	prtbIndexer  cache.Indexer
+	tokenIndexer cache.Indexer
+	authConfigs  v3.AuthConfigInterface
+	secrets      corev1.SecretInterface
+	certs        string
+	caPool       *x509.CertPool
+}
+
+// MigrateActiveDirectoryDNToGUID will go through all Rancher users and check to see if the principalID
+// is an LDAP distinguished name, which was the way we used to map Rancher users to their LDAP entries.
+// If a principalID is based on DN, this will update the user's principalID along with the tokens,
+// CRTBs, and PRTBs to use a principalID that is the objectGUID for the user.
+func MigrateActiveDirectoryDNToGUID(ctx context.Context, management *config.ManagementContext) {
+	m := adMigration{
+		users:        management.Management.Users(""),
+		userLister:   management.Management.Users("").Controller().Lister(),
+		tokenLister:  management.Management.Tokens("").Controller().Lister(),
+		authConfigs:  management.Management.AuthConfigs(""),
+		tokens:       management.Management.Tokens(""),
+		crtbs:        management.Management.ClusterRoleTemplateBindings(""),
+		prtbs:        management.Management.ProjectRoleTemplateBindings(""),
+		crtbIndexer:  management.Management.ClusterRoleTemplateBindings("").Controller().Informer().GetIndexer(),
+		prtbIndexer:  management.Management.ProjectRoleTemplateBindings("").Controller().Informer().GetIndexer(),
+		tokenIndexer: management.Management.Tokens("").Controller().Informer().GetIndexer(),
+		secrets:      management.Core.Secrets(""),
+	}
+
+	migrateCtx, migrateCancel := context.WithCancel(ctx)
+	go func(context.Context, context.CancelFunc) {
+		wait.PollImmediate(time.Hour*24, 0, func() (bool, error) {
+			logrus.Debugf("Starting active directory principalID migration with exponentialBackoff")
+			steps := 5
+			backOffDuration := time.Minute * 10
+			factor := 2
+			var err error
+			for steps > 0 {
+				err = m.migrate()
+				if err != nil {
+					time.Sleep(backOffDuration)
+					backOffDuration = time.Duration(factor) * backOffDuration
+				} else {
+					break
+				}
+				steps--
+			}
+			if err != nil {
+				// returning false & nil because PollImmediate terminates on error
+				logrus.Error(err)
+				return false, nil
+			}
+			// no error returned, user cleanup done, calling the child context's cancelfunc to terminate child context
+			migrateCancel()
```

Comment:
Should this be called in `defer` sooner so that the context is cancelled regardless?